### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/PrimeSpectrum/*) : the collection of minimal prime ideals of a Noetherian ring is finite

### PR DESCRIPTION
**Pull request by jjaassoonn**
Co-PR :  #9087 (maximal ideals of Artinian ring are finite)

Co-authored-by: Andrew Yang <the.erd.one@gmail.com>
Co-authored-by: Junyan Xu <junyanxumath@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Noetherian.lean
Comment (jjaassoonn) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Noetherian.lean:115]: Maybe belongs to a different file, say `RingTheory` somewhere, but then that file needs to import `AlgebraicGeometry`

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:936]: I think something like `orderEquivIrreducibleCloseds` would be a better name (there is [TopologicalSpace.Closeds](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Sets/Closeds.html#TopologicalSpace.Closeds) that you may use).

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:936]: Is irreducibleSetEquivPoints relevant here?

Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:936]: Nice find. On first glance, it seems the order preserving is in the other direction. I will inspect the function more closely later.

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:958]: I think something like `zero_locus ∘ coe '' minimal_primes R = irreducible_components (prime_spectrum R)` or `vanishing_ideal '' irreducible_components (prime_spectrum R) = minimal_primes R` could be useful.

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Noetherian.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Noetherian.lean:115]: Probably we should put the prime spectrum folder under ring theory instead?

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:958]: Moogle found [irreducibleComponents_eq_maximals_closed](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Irreducible.html#irreducibleComponents_eq_maximals_closed). So this should follow from a general construction that an antitone OrderIso sends `minimals` to `maximals`.

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean]: I have changed it to use `irreducibleSetEquivPoints`, thanks!

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean]: the equality has been added

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:953]: I'll add some lemma/defs about OrderEmbedding/Iso and minimals/maximals to make this easy to construct using `OrderIso.setCongr`, `OrderIso.dual` together with [irreducibleComponents_eq_maximals_closed](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Irreducible.html#irreducibleComponents_eq_maximals_closed)

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:951]: With #9190 this becomes
```
  let e : {p : Ideal R | p.IsPrime ∧ ⊥ ≤ p} ≃o PrimeSpectrum R :=
    ⟨⟨fun x ↦ ⟨x.1, x.2.1⟩, fun x ↦ ⟨x.1, x.2, bot_le⟩, fun _ ↦ rfl, fun _ ↦ rfl⟩, Iff.rfl⟩
  (e.trans <| PrimeSpectrum.pointsEquivIrreducibleCloseds R).minimalsIsoMaximals.trans
    (OrderIso.setCongr _ _ <| by simp_rw [irreducibleComponents_eq_maximals_closed, and_comm]).dual
```

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:981]: It might be more useful to show:
+ The vanishing ideal of a closed set is a minimal prime iff the closed set is an irreducible component
+ The zero locus of an ideal is an irreducible component iff the ideal is a minimal prime
and then you can derive this easily.

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:989]: > With #9190 this becomes
> 
> ```
>   let e : {p : Ideal R | p.IsPrime ∧ ⊥ ≤ p} ≃o PrimeSpectrum R :=
>     ⟨⟨fun x ↦ ⟨x.1, x.2.1⟩, fun x ↦ ⟨x.1, x.2, bot_le⟩, fun _ ↦ rfl, fun _ ↦ rfl⟩, Iff.rfl⟩
>   (e.trans <| PrimeSpectrum.pointsEquivIrreducibleCloseds R).minimalsIsoMaximals.trans
>     (OrderIso.setCongr _ _ <| by simp_rw [irreducibleComponent
> ```

Done, thanks for the suggestion!

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:998]: >     * The zero locus of an ideal is an irreducible component iff the ideal is a minimal prime
I could only show that the zero locus of an ideal `I` is an irreducible component iff there exists a minimal ideal `J` with the same zero locus as `I`

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:1021]: Oh yes, radicality is necessary. Thanks!

Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:1021]: The assumption of radicalness and closedness make the two lemmas hard to apply in this equality lemma.

**Comment by erdOne** [comment]
I agree with Junyan that we should use #9190 to show these.
Something like 
```lean
lemma minimalPrimes_eq_minimals {R} [CommRing R] :
    minimalPrimes R = minimals (· ≤ ·) (setOf Ideal.IsPrime) :=
  congr_arg (minimals (· ≤ ·)) (by simp)

lemma PrimeSpectrum.vanishingIdeal_isclosed_isIrreducible {R} [CommRing R] :
    vanishingIdeal (R := R) '' {s | IsClosed s ∧ IsIrreducible s} = setOf Ideal.IsPrime :=
  Set.ext fun I ↦ ⟨fun ⟨s, ⟨hs, hs'⟩, e⟩ ↦
    e ▸ (isIrreducible_zeroLocus_iff_of_radical _ (isRadical_vanishingIdeal s)).mp
      (by rwa [zeroLocus_vanishingIdeal_eq_closure, hs.closure_eq]),
    fun h ↦ ⟨zeroLocus I, ⟨isClosed_zeroLocus I,
      (isIrreducible_zeroLocus_iff_of_radical _ h.isRadical).mpr h⟩,
        (vanishingIdeal_zeroLocus_eq_radical I).trans h.radical⟩⟩

open PrimeSpectrum in
lemma minimalPrimes.eq_irreducibleComponents :
    minimalPrimes R =
    PrimeSpectrum.vanishingIdeal '' (irreducibleComponents <| PrimeSpectrum R) := by
  rw [irreducibleComponents_eq_maximals_closed, minimalPrimes_eq_minimals, ← minimals_swap,
    ← PrimeSpectrum.vanishingIdeal_isclosed_isIrreducible, image_minimals_of_rel_iff_rel]
  intro s t hs _
  exact vanishingIdeal_anti_mono_iff hs.1

lemma PrimeSpectrum.zeroLocus_minimalPrimes :
  zeroLocus ∘ (↑) '' minimalPrimes R =
    irreducibleComponents (PrimeSpectrum R) := by
  rw [minimalPrimes.eq_irreducibleComponents, ← Set.image_comp, Set.EqOn.image_eq_self]
  intros s hs
  simpa [zeroLocus_vanishingIdeal_eq_closure, closure_eq_iff_isClosed]
    using isClosed_of_mem_irreducibleComponents s hs

lemma PrimeSpectrum.mem_irreducibleComponents_iff_vanishingIdeal_mem_minimalPrimes'
    {s : Set (PrimeSpectrum R)} (hs : IsClosed s) :
    s ∈ irreducibleComponents (PrimeSpectrum R) ↔ (vanishingIdeal s) ∈ minimalPrimes R := by
  constructor
  · rw [minimalPrimes.eq_irreducibleComponents]; exact Set.mem_image_of_mem _
  · conv_rhs => rw [← zeroLocus_minimalPrimes, ← hs.closure_eq,
      ← zeroLocus_vanishingIdeal_eq_closure]
    exact Set.mem_image_of_mem _

lemma PrimeSpectrum.zeroLocus_ideal_mem_irreducibleComponents_iff_mem_minimalPrime' (I : Ideal R) :
    zeroLocus I ∈ irreducibleComponents (PrimeSpectrum R) ↔
    ∃ (J : Ideal R), J ∈ minimalPrimes R ∧ zeroLocus I = zeroLocus (J : Set R) := by
  rw [← PrimeSpectrum.zeroLocus_minimalPrimes]
  simp_rw [eq_comm (a := zeroLocus (I : Set R))]
  rfl

lemma PrimeSpectrum.zeroLocus_ideal_mem_irreducibleComponents_iff_mem_minimalPrime_of_radical'
    {I : Ideal R} (hI : I.IsRadical) :
    zeroLocus I ∈ irreducibleComponents (PrimeSpectrum R) ↔ I ∈ minimalPrimes R := by
  constructor
  · conv_rhs => rw [minimalPrimes.eq_irreducibleComponents, ← hI.radical,
      ← vanishingIdeal_zeroLocus_eq_radical]
    exact Set.mem_image_of_mem _
  · rw [← PrimeSpectrum.zeroLocus_minimalPrimes]
    exact Set.mem_image_of_mem _
```

**Comment by jjaassoonn** [comment]
> I agree with Junyan that we should use #9190 to show these. Something like
> 
> ```lean
> lemma minimalPrimes_eq_minimals {R} [CommRing R] :
>     minimalPrimes R = minimals (· ≤ ·) (setOf Ideal.IsPrime) :=
>   congr_arg (minimals (· ≤ ·)) (by simp)
> 
> lemma PrimeSpectrum.vanishingIdeal_isclosed_isIrreducible {R} [CommRing R] :
>     vanishingIdeal (R := R) '' {s | IsClosed s ∧ IsIrreducible s} = setOf Ideal.IsPrime :=
>   Set.ext fun I ↦ ⟨fun ⟨s, ⟨hs, hs'⟩, e⟩ ↦
>     e ▸ (isIrreducible_zeroLocus_iff_of_radical _ (isRadical_vanishingIdeal s)).mp
>       (by rwa [zeroLocus_vanishingIdeal_eq_closure, hs.closure_eq]),
>     fun h ↦ ⟨zeroLocus I, ⟨isClosed_zeroLocus I,
>       (isIrreducible_zeroLocus_iff_of_radical _ h.isRadical).mpr h⟩,
>         (vanishingIdeal_zeroLocus_eq_radical I).trans h.radical⟩⟩
> 
> open PrimeSpectrum in
> lemma minimalPrimes.eq_irreducibleComponents :
>     minimalPrimes R =
>     PrimeSpectrum.vanishingIdeal '' (irreducibleComponents <| PrimeSpectrum R) := by
>   rw [irreducibleComponents_eq_maximals_closed, minimalPrimes_eq_minimals, ← minimals_swap,
>     ← PrimeSpectrum.vanishingIdeal_isclosed_isIrreducible, image_minimals_of_rel_iff_rel]
>   intro s t hs _
>   exact vanishingIdeal_anti_mono_iff hs.1
> 
> lemma PrimeSpectrum.zeroLocus_minimalPrimes :
>   zeroLocus ∘ (↑) '' minimalPrimes R =
>     irreducibleComponents (PrimeSpectrum R) := by
>   rw [minimalPrimes.eq_irreducibleComponents, ← Set.image_comp, Set.EqOn.image_eq_self]
>   intros s hs
>   simpa [zeroLocus_vanishingIdeal_eq_closure, closure_eq_iff_isClosed]
>     using isClosed_of_mem_irreducibleComponents s hs
> 
> lemma PrimeSpectrum.mem_irreducibleComponents_iff_vanishingIdeal_mem_minimalPrimes'
>     {s : Set (PrimeSpectrum R)} (hs : IsClosed s) :
>     s ∈ irreducibleComponents (PrimeSpectrum R) ↔ (vanishingIdeal s) ∈ minimalPrimes R := by
>   constructor
>   · rw [minimalPrimes.eq_irreducibleComponents]; exact Set.mem_image_of_mem _
>   · conv_rhs => rw [← zeroLocus_minimalPrimes, ← hs.closure_eq,
>       ← zeroLocus_vanishingIdeal_eq_closure]
>     exact Set.mem_image_of_mem _
> 
> lemma PrimeSpectrum.zeroLocus_ideal_mem_irreducibleComponents_iff_mem_minimalPrime' (I : Ideal R) :
>     zeroLocus I ∈ irreducibleComponents (PrimeSpectrum R) ↔
>     ∃ (J : Ideal R), J ∈ minimalPrimes R ∧ zeroLocus I = zeroLocus (J : Set R) := by
>   rw [← PrimeSpectrum.zeroLocus_minimalPrimes]
>   simp_rw [eq_comm (a := zeroLocus (I : Set R))]
>   rfl
> 
> lemma PrimeSpectrum.zeroLocus_ideal_mem_irreducibleComponents_iff_mem_minimalPrime_of_radical'
>     {I : Ideal R} (hI : I.IsRadical) :
>     zeroLocus I ∈ irreducibleComponents (PrimeSpectrum R) ↔ I ∈ minimalPrimes R := by
>   constructor
>   · conv_rhs => rw [minimalPrimes.eq_irreducibleComponents, ← hI.radical,
>       ← vanishingIdeal_zeroLocus_eq_radical]
>     exact Set.mem_image_of_mem _
>   · rw [← PrimeSpectrum.zeroLocus_minimalPrimes]
>     exact Set.mem_image_of_mem _
> ```

Thanks! I have incoporated your code into this PR

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:1010]: I think these should be removed in favor of 
```lean
lemma vanishingIdeal_mem_minimalPrimes
    {s : Set (PrimeSpectrum R)} :
    (vanishingIdeal s) ∈ minimalPrimes R ↔ closure s ∈ irreducibleComponents (PrimeSpectrum R) := by
  constructor
  · conv_rhs => rw [← zeroLocus_minimalPrimes,
      ← zeroLocus_vanishingIdeal_eq_closure]
    exact Set.mem_image_of_mem _
  · rw [minimalPrimes.eq_irreducibleComponents, ← vanishingIdeal_closure]
    exact Set.mem_image_of_mem _

lemma zeroLocus_ideal_mem_irreducibleComponents
    {I : Ideal R} :
    zeroLocus I ∈ irreducibleComponents (PrimeSpectrum R) ↔ I.radical ∈ minimalPrimes R := by
  constructor
  · conv_rhs => rw [minimalPrimes.eq_irreducibleComponents,
      ← vanishingIdeal_zeroLocus_eq_radical]
    exact Set.mem_image_of_mem _
  · rw [← PrimeSpectrum.zeroLocus_minimalPrimes, ← zeroLocus_radical]
    exact Set.mem_image_of_mem _
```
maybe even simp lemmas?

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean]: They get rid of `vanishingIdeal` and `zeroLocus` though. Anyway I don't have a strong preference and I'm fine with either way.

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:969]: Following the convention of putting the more complex term on the LHS, maybe we should flip this and call it `PrimeSpectrum.vanishingIdeal_irreducibleComponents`?

Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean]: done

**Comment by erdOne** [comment]
This PR looks good to me now. I rewrote a big chunk of it though so I'll wait for other maintainers (maybe @alreadydone ?) to merge it.

**Review by alreadydone — APPROVED** [review]
Thanks @erdOne for switching to use #9190 and suggesting the better forms of the iff lemmas. LGTM now! I'm not a maintainer😅 but let's send this to maintainers after CI.

**Comment by alreadydone** [comment]
maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**
